### PR TITLE
fix(schedulers): fix asap and animationFrame schedulers to execute across their respective async boundaries

### LINKED ISSUE #1814: Asap and AnimationFrame Schedulers execute synchronously
**RxJS version:**
All

**Code to reproduce:**
<details>
<summary>Expand to see the example code or [run it in esnextb.in here](https://esnextb.in/?gist=2054a372fef49530ace32c76ae79df5f)</summary>
```es6
import { Scheduler } from 'rxjs';

const root = getRootDiv();
const rect = { x: 0, y: 0, w: 200, h: 200};
const box1 = root.appendChild(createBox());
const box2 = root.appendChild(createBox());
const scheduler = Scheduler.animationFrame;

scheduler.schedule(moveBox(box1), 0, {
    ...rect,
    tx: 0, ty: 0,
    ix: rect.w / (rect.w * 2),
    iy: rect.h / (rect.h * 2),
});

scheduler.schedule(moveBox(box2), 0, {
    tx: 0, ty: 0,
    x: rect.w, y: rect.h,
    ix: -1 * rect.w / (rect.w * 2),
    iy: -1 * rect.h / (rect.h * 2),
});

function moveBox(box) {
    return function(state) {
        const { x, y, tx, ty } = state;
        if (tx < rect.w && ty < rect.h) {
            box.style.transform = `translate3d(${x}px, ${y}px, 0px)`;
            state.x = x + state.ix;
            state.y = y + state.iy;
            state.tx = tx + Math.abs(state.ix);
            state.ty = ty + Math.abs(state.iy);
            this.schedule(state);
        }
    }
}

function createBox(parent) {
  const box = document.createElement('div');
  box.style.width = `${rect.w}px`;
  box.style.height = `${rect.h}px`;
  box.style.opacity = 0.25;
  box.style.background = '#000000';
  box.style.position = 'absolute';
  return box;
}

function getRootDiv(root) {
  root = (
    document.getElementById('root') ||
    document.body.appendChild((
      root = document.createElement('div')) && (
      root.id = 'root') && (
      root)
    )
  );
  root.style.top = 0;
  root.style.left = 0;
  root.style.right = 0;
  root.style.bottom = 0;
  root.style.overflow = 'hidden';
  root.style.position = 'absolute';
  root.style['transform-style'] = 'preserve-3d';
  return root;
}
```
</details>

**Expected behavior:**
The boxes should animate by repeated calls to `setImmediate` or `requestAnimationFrame`.

**Actual behavior:**
They don't. Instead, one call to `setImmediate` or `requestAnimationFrame` happens, then the actions reschedule themselves synchronously.

**Additional information:**
I've fixed this in a [branch](https://github.com/trxcllnt/RxJS/tree/scheduler-action-fixes) and will have a PR once I've worked out the performance regressions.

### PR DESCRIPTION
**Pull request by trxcllnt**
**Description:**

The AsapScheduler and AnimationFrameScheduler were totally busted. My bad. Now they execute their scheduled actions in batches. If actions reschedule while executing a batch, a new frame is requested for the rescheduled action to execute in.

Other updates included in this PR:

* simplifies the public `Scheduler` and `Action` APIs. Implementation details like the `actions` queue and `active` boolean are now on the concrete implementations, so it's easier for people to implement the Scheduler API.
* renames `FutureAction` -> `AsyncAction` to conform to the same naming convention as the rest of the Action types.
* Actions should be able to be recursively rescheduled with different delay times and switch between their delay = 0 and delay > 0 behaviors. For example, a delay=0 `QueueAction` should be able to reschedule with delay > 0 and fall back to its parent class's (`AsyncAction`) behavior.
* If a scheduled action throws an error, the schedulers now dispose (and un-reference) the rest of the scheduled actions before re-throwing. I'm not sure how necessary this is, but it did seem like keeping the actions in the queue could was incorrect (they'd stick around until the Scheduler was flushed, at which time they'd be executed) or possibly leaky.

**Related issue (if exists):**

Fixes #1814

**Performance results:**

Re-ran a few of the performance tests to check regressions when using the queue scheduler. No discernible perf hits.

Current master:

                                         |                     RxJS 4.1.0 |             RxJS 5.0.0-beta.10 |          factor |      % improved
-----------------------------------------|--------------------------------|--------------------------------|-----------------|----------------
                      concat - immediate |                50,291 (±1.21%) |               357,051 (±2.00%) |           7.10x |          610.0%
                       merge - immediate |                 5,621 (±1.53%) |                56,834 (±1.00%) |          10.11x |          911.1%
                                  concat |                21,527 (±1.74%) |               153,550 (±4.04%) |           7.13x |          613.3%
                                   merge |                 2,283 (±2.35%) |                21,295 (±2.45%) |           9.33x |          832.8%
             from-with-array - immediate |               127,203 (±1.81%) |               895,727 (±1.10%) |           7.04x |          604.2%
          from-with-iterable - immediate |               105,449 (±2.19%) |               357,870 (±3.85%) |           3.39x |          239.4%
            from-with-string - immediate |                60,665 (±4.35%) |               384,754 (±3.69%) |           6.34x |          534.2%
                          of - immediate |                59,050 (±3.56%) |               829,434 (±3.29%) |          14.05x |        1,304.6%
                         from-with-array |                54,174 (±3.47%) |               668,259 (±3.17%) |          12.34x |        1,133.6%
                      from-with-iterable |                51,085 (±3.41%) |               366,127 (±2.51%) |           7.17x |          616.7%
                        from-with-string |                33,901 (±2.78%) |               365,895 (±3.86%) |          10.79x |          979.3%
                                      of |                60,350 (±3.36%) |               443,226 (±2.38%) |           7.34x |          634.4%

This PR:

                                         |                     RxJS 4.1.0 |             RxJS 5.0.0-beta.10 |          factor |      % improved
-----------------------------------------|--------------------------------|--------------------------------|-----------------|----------------
                      concat - immediate |                54,293 (±1.61%) |               386,044 (±1.16%) |           7.11x |          611.0%
                       merge - immediate |                 5,751 (±1.11%) |                57,951 (±0.86%) |          10.08x |          907.6%
                                  concat |                22,460 (±1.07%) |               169,922 (±1.51%) |           7.57x |          656.5%
                                   merge |                 2,415 (±1.21%) |                22,674 (±1.01%) |           9.39x |          838.7%
             from-with-array - immediate |               128,302 (±1.54%) |               851,673 (±1.70%) |           6.64x |          563.8%
          from-with-iterable - immediate |               114,602 (±1.34%) |               364,576 (±1.75%) |           3.18x |          218.1%
            from-with-string - immediate |                60,711 (±4.18%) |               367,734 (±3.85%) |           6.06x |          505.7%
                          of - immediate |                58,137 (±3.61%) |               853,370 (±3.29%) |          14.68x |        1,367.9%
                         from-with-array |                51,498 (±3.14%) |               691,779 (±3.70%) |          13.43x |        1,243.3%
                      from-with-iterable |                51,048 (±3.52%) |               372,203 (±3.40%) |           7.29x |          629.1%
                        from-with-string |                33,374 (±3.06%) |               394,613 (±2.56%) |          11.82x |        1,082.4%
                                      of |                53,013 (±4.33%) |               395,391 (±2.94%) |           7.46x |          645.8%

Side-by-side:

                                         |    RxJS 5.0.0-beta.10 (master) |        RxJS 5.0.0-beta.10 (PR)
-----------------------------------------|--------------------------------|-------------------------------
                      concat - immediate |               357,051 (±2.00%) |               386,044 (±1.16%)
                       merge - immediate |                56,834 (±1.00%) |                57,951 (±0.86%)
                                  concat |               153,550 (±4.04%) |               169,922 (±1.51%)
                                   merge |                21,295 (±2.45%) |                22,674 (±1.01%)
             from-with-array - immediate |               895,727 (±1.10%) |               851,673 (±1.70%)
          from-with-iterable - immediate |               357,870 (±3.85%) |               364,576 (±1.75%)
            from-with-string - immediate |               384,754 (±3.69%) |               367,734 (±3.85%)
                          of - immediate |               829,434 (±3.29%) |               853,370 (±3.29%)
                         from-with-array |               668,259 (±3.17%) |               691,779 (±3.70%)
                      from-with-iterable |               366,127 (±2.51%) |               372,203 (±3.40%)
                        from-with-string |               365,895 (±3.86%) |               394,613 (±2.56%)
                                      of |               443,226 (±2.38%) |               395,391 (±2.94%)

### CODE REVIEW

File: src/Scheduler.ts [hunk 1]
```diff
@@ -1,11 +1,59 @@
...
+ *
+ * ```ts
+ * class Scheduler {
+ *   now(): number;
+ *   schedule(work, delay?, state?): Subscription;
+ * }
+ * ```
+ *
+ * @class Scheduler
+ */
```
Comment (staltz) [inline src/Scheduler.ts]: For anyone else reviewing this PR: JSDoc in this file looks correct and should render correctly on the site. 👍

File: src/scheduler/MiscJSDoc.ts [hunk 2]
```diff
@@ -1,85 +0,0 @@
...
- *   execute(): void;
- *   scheduler: Scheduler;
- *   error: any;
- * }
- * ```
- *
- * @interface
- * @name Action
- * @noimport true
- */
```
Comment (staltz) [inline src/scheduler/MiscJSDoc.ts]: Also the removal of this file and moving the JSDoc to the Action class seems correct for the docs site.

**Comment by blesh** [comment]
LGTM :shipit:

**Comment by blesh** [comment]
lgtm

**Comment by trxcllnt** [comment]
@blesh what's the issue?

File: src/scheduler/MiscJSDoc.ts (hunk 2 above)
Comment (blesh) [inline src/scheduler/MiscJSDoc.ts]: @trxcllnt the only issue was right here. Pretty trivial

Comment (trxcllnt) [inline src/scheduler/MiscJSDoc.ts]: @blesh from what I can tell, @staltz is saying this change is correct, right?

Comment (staltz) [inline src/scheduler/MiscJSDoc.ts]: Yeah it's correct. I just commented so others can ignore these changes, they're already verified and LGTM. I didn't review carefully the whole PR so I didn't give my LGTM yet

File: src/Rx.ts [hunk 3]
```diff
@@ -184,6 +184,8 @@ import observable from 'symbol-observable';
  * asynchronous conversions.
  * @property {Scheduler} async Schedules work with `setInterval`. Use this for
  * time-based operations.
+ * @property {Scheduler} animation Schedules work with `requestAnimationFrame`.
```
Comment (saneyuki) [inline src/Rx.ts]: should this `animation` be `animationFrame`?

**Comment by blesh** [comment]
Looks like this needs a rebase, @trxcllnt I'd like to get it merged.

**Comment by trxcllnt** [comment]
@blesh rebased

File: src/Rx.ts (hunk 3 above)
Comment (blesh) [inline src/Rx.ts]: Oops, yup @trxcllnt this needs fixed.

File: src/scheduler/Action.ts [hunk 4]
```diff
@@ -1,12 +1,35 @@
...
+export class Action<T> extends Subscription {
+  constructor(scheduler: Scheduler, work: (state?: T) => void) {
+    super();
+  }
+  /**
+   * Schedules this action on its parent Scheduler for execution. May be passed
+   * some context object, `state`. May happen at some point in the future,
+   * according to the `delay` parameter, if specified.
+   * @param {any} [state] Some contextual data that the `work` function uses when
+   * called by the Scheduler.
```
Comment (saneyuki) [inline src/scheduler/Action.ts]: Is this `{any}` jsdoc correct?  Should it be `{T}` with `@template T`?

Comment (trxcllnt) [inline src/scheduler/Action.ts]: @saneyuki I think you're right. I'll update the PR.

File: src/Rx.ts (hunk 3 above)
Comment (trxcllnt) [inline src/Rx.ts]: @blesh ok, I'll update. Though I was thinking `Scheduler.animation` vs. `Scheduler.animationFrame` would be a nice alias.

**Pull request merged**